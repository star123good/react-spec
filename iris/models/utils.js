// @flow
const debug = require('debug')('iris:models:utils');
import processChangefeed from 'rethinkdb-changefeed-reconnect';
import { db } from './db';
import Raven from 'shared/raven';
import type { Cursor } from 'rethinkdbdash';

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table, cb) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true })
      .then(cursor => {
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
        return cursor;
      })
  );
};

export const createChangefeed = (
  getChangefeed: () => Promise<Cursor>,
  callback: (arg: any) => void,
  name?: string
) => {
  return processChangefeed(
    getChangefeed,
    callback,
    err => {
      console.error(err);
      Raven.captureException(err);
    },
    {
      changefeedName: name,
      attemptDelay: 60000,
      maxAttempts: Infinity,
      logger: {
        // Ignore log and info logs in production
        log: debug,
        info: debug,
        warn: console.warn.bind(console),
        error: console.error.bind(console),
      },
    }
  );
};

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(db.row('lastSeen').during(db.now().sub(current), db.now()))
    .count()
    .default(0)
    .run();
};

export const getGrowth = async (
  table: string,
  range: Timeframe,
  field: string,
  filter: ?mixed
) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = () => {
  return db
    .table('coreMetrics')
    .orderBy('date')
    .run();
};
