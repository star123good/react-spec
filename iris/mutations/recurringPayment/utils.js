// @flow
export const parseStripeErrors = err => {
  switch (err.type) {
    case 'StripeCardError':
      // A declined card error
      return new UserError(err); // => e.g. "Your card's expiration year is invalid."
      break;
    case 'RateLimitError':
      // Too many requests made to the API too quickly
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 1'
      );
      break;
    case 'StripeInvalidRequestError':
      // Invalid parameters were supplied to Stripe's API
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 2'
      );
      break;
    case 'StripeAPIError':
      // An error occurred internally with Stripe's API
      return new UserError(
        'Something went wrong at Stripe, try again later: 3'
      );
      break;
    case 'StripeConnectionError':
      // Some kind of error occurred during the HTTPS communication
      return new UserError(
        'Something went wrong at Stripe, try again later: 4'
      );
      break;
    case 'StripeAuthenticationError':
      // You probably used an incorrect API key
      return new UserError(
        'Something went wrong at Stripe, try again later: 5'
      );
      break;
    default:
      // Handle any other types of unexpected errors
      return new UserError('Something went wrong, try again later: 6');
      break;
  }
};

export const getStripeCustomer = (customerId: string) => {
  try {
    return stripe.customers.retrieve(customerId);
  } catch (err) {
    return console.log(err) || err;
  }
};

export const createStripeCustomer = (email: string, source: string) => {
  try {
    return stripe.customers.create({
      email,
      source,
    });
  } catch (err) {
    return console.log(err) || err;
  }
};

export const updateStripeCustomer = (customer, email, source) => {
  try {
    return stripe.customers.update(customer, {
      email,
      source,
    });
  } catch (err) {
    return console.log(err) || err;
  }
};

export const createStripeSubscription = (
  customer: string,
  plan: string,
  quantity: number
) => {
  try {
    return stripe.subscriptions.create({
      customer,
      items: [
        {
          plan,
          quantity: quantity ? quantity : 1, // if no quantity is provided, default to one
        },
      ],
    });
  } catch (err) {
    return console.log(err) || err;
  }
};

export const deleteStripeSubscription = (subscription: string) => {
  try {
    return stripe.subscriptions.del(subscription);
  } catch (err) {
    return console.log(err) || err;
  }
};
