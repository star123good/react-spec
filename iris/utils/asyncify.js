// @flow
// Turn a callback-based listener into an async iterator
const Queue = require('then-queue');

const asyncify = (
  listener: (...args: any[]) => any,
  onError?: Error => void
) => {
  const queues = [];
  listener((...args) => queues.map(q => q.push(...args)));

  return () => {
    const queue = new Queue();
    queues.push(queue);

    // I have no idea how to type generators...
    // $FlowFixMe
    async function* AsyncGenerator() {
      try {
        while (true) {
          yield await queue.pop();
        }
      } catch (err) {
        onError && onError(err);
      }
    }

    return AsyncGenerator();
  };
};

export default asyncify;
