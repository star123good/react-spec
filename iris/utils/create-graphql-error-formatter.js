import Raven from 'raven';
import { IsUserError } from './UserError';

const createGraphQLErrorFormatter = req => error => {
  const isUserError = error.originalError
    ? error.originalError[IsUserError]
    : false;
  let sentryId = 'ID only generated in production';
  if (!isUserError) {
    console.log(error);
    if (process.env.NODE_ENV === 'production') {
      sentryId = Raven.captureException(error, Raven.parsers.parseRequest(req));
    }
  }
  return {
    message: isUserError ? error.message : `Internal server error: ${sentryId}`,
    // Hide the stack trace in production mode
    stack:
      !process.env.NODE_ENV === 'production' ? error.stack.split('\n') : null,
  };
};

export default createGraphQLErrorFormatter;
