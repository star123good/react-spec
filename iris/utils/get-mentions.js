// @flow
import { MENTIONS } from 'shared/regexps';

export default (text: string): Array<string> => {
  return text.match(MENTIONS) || [];
};
