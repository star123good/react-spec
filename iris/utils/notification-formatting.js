//@flow
import createEmojiRegex from 'emoji-regex';
// This regex matches every string with any emoji in it, not just strings that only have emojis
const originalEmojiRegex = createEmojiRegex();
// Make sure we match strings that only contain emojis (and whitespace)
const regex = new RegExp(
  `^(${originalEmojiRegex.toString().replace(/\/g$/, '')}|\\s)+$`
);
export const onlyContainsEmoji = (text: string) => regex.test(text);

export const getDistinctNotifications = array => {
  let unique = {};
  let distinct = [];
  for (let i in array) {
    if (typeof unique[array[i].id] === 'undefined') {
      distinct.push(array[i]);
    }
    unique[array[i].id] = 0;
  }
  return distinct;
};

export const parseNotification = notification => {
  return Object.assign({}, notification, {
    actors: notification.actors.map(actor => {
      return {
        id: actor.id,
        type: actor.type,
        payload: JSON.parse(actor.payload),
      };
    }),
    context: {
      id: notification.context.id,
      type: notification.context.type,
      payload: JSON.parse(notification.context.payload),
    },
    entities: notification.entities.map(entity => {
      return {
        id: entity.id,
        type: entity.type,
        payload: JSON.parse(entity.payload),
      };
    }),
  });
};

const actorsToString = actors => {
  // reverse to show the most recent first
  const names =
    actors &&
    actors.length > 0 &&
    actors.map(actor => JSON.parse(actor.payload).name).reverse();

  if (actors.length === 1) {
    return names[0];
  } else if (actors.length === 2) {
    return `${names[0]} and ${names[1]}`;
  } else if (actors.length === 3) {
    return `${names[0]}, ${names[1]} and ${names[2]}`;
  } else {
    return `${names[0]} and ${names.length - 1} others`;
  }
};

const actorsToObjects = actors => {
  return (
    actors &&
    actors.length > 0 &&
    actors
      .map(actor => {
        return {
          name: actor.payload.name,
          username: actor.payload.username,
          profilePhoto: actor.payload.profilePhoto,
          id: actor.payload.id,
        };
      })
      .reverse()
  );
};

export const parseActors = (actors, currentUser) => {
  const filteredActors = actors.filter(actor => actor.id !== currentUser.id);
  const asString = actorsToString(filteredActors);
  const asObjects = actorsToObjects(filteredActors);

  return {
    asString,
    asObjects,
  };
};

export const parseEvent = event => {
  switch (event) {
    case 'MESSAGE_CREATED': {
      return 'replied';
    }
    case 'REACTION_CREATED': {
      return 'liked';
    }
    case 'CHANNEL_CREATED': {
      return 'created a channel';
    }
    case 'USER_JOINED_COMMUNITY': {
      return 'joined';
    }
    default: {
      console.log('Not a valid event type');
    }
  }
};

const threadToString = (context, currentUser) => {
  const payload = JSON.parse(context.payload);
  const isCreator = payload.creatorId === currentUser.id;
  const str = isCreator ? 'in your thread' : 'in';
  return `${str} ${payload.content.title}`;
};

const messageToString = () => 'your reply';

const communityToString = context => {
  return `${JSON.parse(context.payload).name}`;
};

const channelToString = context => {
  return `${JSON.parse(context.payload).name}`;
};

export const parseContext = (context, currentUser) => {
  switch (context.type) {
    case 'SLATE':
    case 'THREAD': {
      const asString = threadToString(context, currentUser);
      return {
        asString,
      };
    }
    case 'MESSAGE': {
      const asString = messageToString(context);
      return {
        asString,
      };
    }
    case 'COMMUNITY': {
      const asString = communityToString(context);
      return {
        asString,
      };
    }
    case 'CHANNEL': {
      const asString = channelToString(context);
      return {
        asString,
      };
    }
    default: {
      console.log('Invalid notification context type');
    }
  }
};

export const getLastMessageCreatedByAnotherUser = (entities, currentUser) => {
  const filteredMessages = entities.filter(
    entity => JSON.parse(entity.payload).senderId !== currentUser.id
  );
  return JSON.parse(filteredMessages[filteredMessages.length - 1].payload);
};
