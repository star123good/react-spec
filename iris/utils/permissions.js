// @flow
import {
  COMMUNITY_SLUG_BLACKLIST,
  CHANNEL_SLUG_BLACKLIST,
} from 'shared/slug-blacklists';

export const isAdmin = (id: string): boolean => {
  const admins = [
    'gVk5mYwccUOEKiN5vtOouqroGKo1',
    '01p2A7kDCWUjGj6zQLlMQUOSQL42',
    'VToKcde16dREgDkXcDl3hhcrFN33',
  ];
  return admins.indexOf(id) > -1;
};

export const communitySlugIsBlacklisted = (slug: string): boolean => {
  return COMMUNITY_SLUG_BLACKLIST.indexOf(slug) > -1;
};

export const channelSlugIsBlacklisted = (slug: string): boolean => {
  return CHANNEL_SLUG_BLACKLIST.indexOf(slug) > -1;
};
