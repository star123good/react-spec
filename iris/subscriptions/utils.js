// @flow
import { getUserPermissionsInChannel } from '../models/usersChannels';
import { getChannels } from '../models/channel';
import type { DBThread } from '../models/thread';
import type { GraphQLContext } from '../';

// Check that a user can view a channel
export const userCanViewChannel = async (channelId: string, userId: string) => {
  const [channel] = await getChannels([channelId]);
  // Everybody can view public channels
  if (!channel.isPrivate) return true;

  // If you're not signed in you can't subscribe to private channels
  if (!userId) return false;

  const { isMember } = await getUserPermissionsInChannel(channelId, userId);
  return isMember;
};
