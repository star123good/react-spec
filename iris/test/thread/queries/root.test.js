// @flow
import { request } from '../../utils';

it('should fetch a thread', async () => {
  const query = /* GraphQL */ `
    {
      thread(id: "ce2b4488-4c75-47e0-8ebc-2539c1e6a193") {
        id
        createdAt
        modifiedAt
        lastActive
        isPublished
        isLocked
        type
        content {
          title
        }
      }
    }
  `;

  expect.hasAssertions();
  const result = await request(query);

  expect(result).toMatchSnapshot();
});
