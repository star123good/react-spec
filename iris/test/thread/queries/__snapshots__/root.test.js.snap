// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch a thread 1`] = `
Object {
  "data": Object {
    "thread": Object {
      "content": Object {
        "title": "The first thread! ðŸŽ‰",
      },
      "createdAt": "2016-12-31T23:00:00.000Z",
      "id": "ce2b4488-4c75-47e0-8ebc-2539c1e6a193",
      "isLocked": false,
      "isPublished": true,
      "lastActive": "2016-12-31T23:00:00.000Z",
      "modifiedAt": "2016-12-31T23:00:00.000Z",
      "type": "DRAFTJS",
    },
  },
}
`;
